// @crystallize/node-service-api-request-handlers/dist/payments/adyen/handlers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleAdyenWebhookRequestPayload = exports.handleAdyenPaymentSessionPayload = void 0;
const api_library_1 = require("@adyen/api-library");
async function handleAdyenPaymentSessionPayload(payload, args) {
    const cart = await args.fetchCart();
    const client = new api_library_1.Client({
        apiKey: args.apiKey,
        environment: args.env,
    });
    const checkout = new api_library_1.CheckoutAPI(client);
    const response = await checkout.sessions({
        amount: {
            currency: args.currency,
            value: cart.total.gross * 100,
        },
        countryCode: args.countryCode,
        merchantAccount: args.merchantAccount,
        reference: payload.cartId,
        returnUrl: args.returnUrl,
    });
    if (!response) {
        throw new Error('Session not created');
    }
    return {
        sessionId: response.id,
        data: response,
    };
}
exports.handleAdyenPaymentSessionPayload = handleAdyenPaymentSessionPayload;
async function handleAdyenWebhookRequestPayload(payload, args) {
    return await args.handleEvent(payload);
}
exports.handleAdyenWebhookRequestPayload = handleAdyenWebhookRequestPayload;
