// @crystallize/reactjs-hooks/dist/core/MainProvider/Reducer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapToReducerActions = exports.Reducer = void 0;
function Reducer(state, action) {
    switch (action.type) {
        case 'LOADING': {
            return {
                ...state,
                loading: action.state,
            };
        }
        case 'CHANGE_LANGUAGE': {
            return {
                ...state,
                language: action.language,
            };
        }
        case 'UPDATE_CONFIGURATION': {
            return {
                ...state,
                configuration: action.configuration,
            };
        }
        default: {
            throw new Error('Main Provider - Unhandled action type');
        }
    }
}
exports.Reducer = Reducer;
function mapToReducerActions(dispatch) {
    return {
        loading: (state) => dispatch({ type: 'LOADING', state }),
        changeLanguage: (language) => dispatch({ type: 'CHANGE_LANGUAGE', language }),
        updateConfiguration: (configuration) => dispatch({ type: 'UPDATE_CONFIGURATION', configuration }),
    };
}
exports.mapToReducerActions = mapToReducerActions;
