// @crystallize/node-service-api-request-handlers/dist/payments/stripe/handlers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleStripePaymentIntentWebhookRequestPayload = exports.handleStripeCreatePaymentIntentRequestPayload = void 0;
const stripe_1 = __importDefault(require("stripe"));
async function handleStripeCreatePaymentIntentRequestPayload(payload, args) {
    const stripe = new stripe_1.default(args.secret_key, {
        apiVersion: '2022-08-01',
    });
    const cart = await args.fetchCart();
    const intentArguments = args.createIntentArguments(cart);
    const defaultMetadata = {
        cartId: payload.cartId,
    };
    const paymentIntent = await stripe.paymentIntents.create({
        amount: intentArguments.amount,
        currency: intentArguments.currency,
        automatic_payment_methods: {
            enabled: true,
        },
        metadata: {
            ...defaultMetadata,
            ...(intentArguments.metatdata ? intentArguments.metatdata : {}),
        },
        ...(args.otherIntentArguments ? args.otherIntentArguments : {}),
    });
    return {
        key: paymentIntent.client_secret || '',
    };
}
exports.handleStripeCreatePaymentIntentRequestPayload = handleStripeCreatePaymentIntentRequestPayload;
async function handleStripePaymentIntentWebhookRequestPayload(payload, args) {
    const stripe = new stripe_1.default(args.secret_key, {
        apiVersion: '2022-08-01',
    });
    const event = stripe.webhooks.constructEvent(args.rawBody, args.signature, args.endpointSecret);
    return await args.handleEvent(event.type, event);
}
exports.handleStripePaymentIntentWebhookRequestPayload = handleStripePaymentIntentWebhookRequestPayload;
