// @crystallize/node-service-api-request-handlers/dist/payments/montonio/handlers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleMontonioPaymentUpdateWebhookRequestPayload = exports.handleMontonioCreatePaymentLinkRequestPayload = void 0;
const jsonwebtoken_1 = __importDefault(require("jsonwebtoken"));
async function handleMontonioCreatePaymentLinkRequestPayload(payload, args) {
    const cart = await args.fetchCart();
    const paymentArguments = args.createPaymentArguments(cart);
    const montonioPayload = {
        amount: paymentArguments.amount,
        currency: paymentArguments.currency.toLocaleUpperCase(),
        access_key: args.access_key,
        merchant_reference: payload.cartId,
        merchant_return_url: paymentArguments.urls.return,
        merchant_notification_url: paymentArguments.urls.notification,
        payment_information_unstructured: `Payment for cart ${payload.cartId}`,
        checkout_email: paymentArguments.customer.email,
        checkout_phone_number: paymentArguments.customer.phone,
        checkout_first_name: paymentArguments.customer.firstName,
        checkout_last_name: paymentArguments.customer.lastName,
        ...(args.otherPaymentArgumentsForLink ? args.otherPaymentArgumentsForLink : {}),
    };
    const token = jsonwebtoken_1.default.sign(montonioPayload, args.secret_key, { algorithm: 'HS256', expiresIn: '10m' });
    return { url: `https://${args.origin}?payment_token=${token}` };
}
exports.handleMontonioCreatePaymentLinkRequestPayload = handleMontonioCreatePaymentLinkRequestPayload;
async function handleMontonioPaymentUpdateWebhookRequestPayload(payload, args) {
    const decoded = jsonwebtoken_1.default.verify(args.token, args.secret_key);
    return await args.handleEvent(decoded);
}
exports.handleMontonioPaymentUpdateWebhookRequestPayload = handleMontonioPaymentUpdateWebhookRequestPayload;
